runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,354 @@
+                    .increment_total_prioritization_fee(priority_details.priority);
+            },
+            "update_time",
+        );
+
+        self.metrics
+            .increment_total_update_elapsed_us(update_time.as_us());
+        Ok(())
+    }
+
+    /// Accounts that have minimum fees lesser or equal to the minimum fee in the block are redundant, they are
+    /// removed to reduce memory footprint.
+    pub fn prune_irrelevant_writable_accounts(&mut self) {
+        self.metrics.total_writable_accounts_count = self.get_writable_accounts_count() as u64;
+        self.min_writable_account_fees
+            .retain(|_, account_fee| account_fee > &mut self.min_transaction_fee);
+        self.metrics.relevant_writable_accounts_count = self.get_writable_accounts_count() as u64;
+    }
+
+    pub fn mark_block_completed(&mut self) -> Result<(), PrioritizationFeeError> {
+        if self.is_finalized {
+            return Err(PrioritizationFeeError::BlockIsAlreadyFinalized);
+        }
+        self.is_finalized = true;
+        Ok(())
+    }
+
+    pub fn get_min_transaction_fee(&self) -> Option<u64> {
+        (self.min_transaction_fee != u64::MAX).then(|| self.min_transaction_fee)
+    }
+
+    pub fn get_writable_account_fee(&self, key: &Pubkey) -> Option<u64> {
+        self.min_writable_account_fees.get(key).copied()
+    }
+
+    pub fn get_writable_account_fees(&self) -> impl Iterator<Item = (&Pubkey, &u64)> {
+        self.min_writable_account_fees.iter()
+    }
+
+    pub fn get_writable_accounts_count(&self) -> usize {
+        self.min_writable_account_fees.len()
+    }
+
+    pub fn is_finalized(&self) -> bool {
+        self.is_finalized
+    }
+
+    pub fn report_metrics(&self, slot: Slot) {
+        self.metrics.report(slot);
+
+        // report this slot's min_transaction_fee and top 10 min_writable_account_fees
+        let min_transaction_fee = self.get_min_transaction_fee().unwrap_or(0);
+        let mut accounts_fees: Vec<_> = self.get_writable_account_fees().collect();
+        accounts_fees.sort_by(|lh, rh| rh.1.cmp(lh.1));
+        datapoint_info!(
+            "block_min_prioritization_fee",
+            ("slot", slot as i64, i64),
+            ("entity", "block", String),
+            ("min_prioritization_fee", min_transaction_fee as i64, i64),
+        );
+        for (account_key, fee) in accounts_fees.iter().take(10) {
+            datapoint_info!(
+                "block_min_prioritization_fee",
+                ("slot", slot as i64, i64),
+                ("entity", account_key.to_string(), String),
+                ("min_prioritization_fee", **fee as i64, i64),
+            );
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        solana_sdk::{
+            compute_budget::ComputeBudgetInstruction, message::Message, pubkey::Pubkey,
+            system_instruction, transaction::Transaction,
+        },
+    };
+
+    fn build_sanitized_transaction_for_test(
```

Comment:
Seems like it would be worthwhile to test briefly on transactions with no write_account, and also with several. Could fold that into your primary test_update_prioritization_fee test.